File: leetcode/C++/super-palindromes.cpp
class Solution {
public:
    bool palinedrome(int &s, int &e, long long num){
        if(num<s || num>e) return false;
        num*=num;
        string n=to_string(num);
        int l=n.length()-1;
        for(int i=0; i<l/2; ++i) if(n[i]!=n[l-i]) return false;
        return true;
    }
    int superpalindromesInRange(string left, string right) {
        int arr[]={0,1,10,100,1000,10000,100000};
        long long n1=stoll(left),n2=stoll(right);
        int s=sqrt(n1), e=sqrt(n2), sol=0;
        string t,r;
        for(int i=1; i<4; ++i) if(s<=i && i<=e) ++sol;
        for(int i=to_string(s).length(); i<=to_string(e).length(); ++i){
            for(int j=arr[i/2]; j<arr[i/2+1]; ++j){
                t=r=to_string(j); reverse(r.begin(),r.end());
                if(i%2){
                    for(char k='0'; k<='9'; ++k) if(palinedrome(s,e,stoll(t+k+r))) ++sol;
                }
                else {
                    if(palinedrome(s,e,stoll(t+r))) ++sol;
                }
            }
        }
        return sol;
    }
};

File: leetcode/Python/super-palindromes.py
class Solution:
    def superpalindromesInRange(self, left: str, right: str) -> int:
        arr=[0,1,10,100,1000,10000,100000]
        c=[str(i) for i in range(10)]
        sol, t, rs, re = 0, 0, int(sqrt(int(left))), int(sqrt(int(right)))
        for i in range(1,4):
            if rs<=i<=re:
                sol+=1
        for i in range(len(str(rs)),len(str(re))+1):
            m=i//2
            for j in range(arr[m],arr[m+1]):
                if i%2 > 0:
                    for k in c:
                        t=int(str(j)+k+str(j)[::-1])
                        if rs<=t<=re:
                            t*=t
                            if str(t) == str(t)[::-1]:
                                sol+=1
                else:
                    t=int(str(j)+str(j)[::-1])
                    if rs<=t<=re:
                        t*=t
                        if str(t) == str(t)[::-1]:
                            sol+=1
        return sol